ethcore/light/src/on_demand/request.rs

```diff
@@ -26,17 +26,383 @@ use ethcore::receipt::Receipt;
 use ethcore::state::{self, ProvedExecution};
 use ethcore::transaction::SignedTransaction;
 
+use request::{self as net_request, IncompleteRequest, Output, OutputKind};
+
 use rlp::{RlpStream, UntrustedRlp};
-use util::{Address, Bytes, DBValue, HashDB, H256, U256};
+use util::{Address, Bytes, DBValue, HashDB, Mutex, H256, U256};
 use util::memorydb::MemoryDB;
 use util::sha3::Hashable;
 use util::trie::{Trie, TrieDB, TrieError};
 
+const SUPPLIED_MATCHES: &'static str = "supplied responses always match produced requests; qed";
```

Comment:
https://github.com/paritytech/parity/pull/5419/files/0fd3e36c23123ddd7f33b3891979214e3f23ceb0#diff-abb081d3b17a1ecf465e845b3fef0728R354

this match ensures that a response is invalid if it doesn't have the correct type for the request.